File: pranav_LLD_Aug/Behavioral_design_pattern/Observer_design_pattern/WeatherStation.java
import java.util.ArrayList;
import java.util.List;

public class WeatherStation implements ISubject {

    List<IObserver> observerList;

    public WeatherStation() {
        observerList = new ArrayList<>();
    }

    @Override
    public void addObserver(IObserver observer) {
        observerList.add(observer);
    }

    @Override
    public void removeObserver(IObserver observer) {
        observerList.remove(observer);
    }

    @Override
    public void notifyObservers(WeatherEvent weatherEvent) {

        System.out.println("Publishing message to observers");
        for(IObserver observer : observerList) {
            observer.update(weatherEvent);
        }

    }

    public void display() {
        for (IObserver o : observerList) {
            o.display();
        }
    }
}

File: rahul-narang-LLD-March/src/WeatherStation.kt
class WeatherStation: WeatherStationSubject {

    private val observers = mutableListOf<Observer>()

    private lateinit var weatherStationMessage: WeatherStationMessage

    fun setWeatherData(weatherStationMessage: WeatherStationMessage) {
        this.weatherStationMessage = weatherStationMessage
        notifyObservers(weatherStationMessage)
    }

    override fun registerObserver(observer: Observer) {
        observers.add(observer)
    }

    override fun removeObserver(observer: Observer) {
        observers.remove(observer)
    }

    override fun notifyObservers(weatherStationMessage: WeatherStationMessage) {
        observers.forEach {
            it.update(weatherStationMessage)
        }
    }
}